Complete Mail Client Implementation with FastAPI Service

ranjanp26 opened the pull request:
# Pull Request

## Summary
- **Problem being solved**: Implement a complete mail client system with Gmail API integration and FastAPI service layer
- **Why now**: Assignment submission - demonstrates component-based architecture, dependency injection, and modern Python development practices
- **User-visible impact**: Fully functional mail client accessible via direct Gmail API or REST API endpoints, with Docker support for easy deployment


## Change Type
- [x] ✨ Feature / enhancement
- [x] 📚 Documentation
- [x] 🧪 Test improvement
- [x] 🔨 Build / tooling

## Impacted Areas
- [x] `mail_client_api` - Abstract contracts and interfaces
- [x] `gmail_client_impl` - Gmail API implementation
- [x] `mail_client_service` - **NEW**: FastAPI service wrapper
- [x] Documentation - Comprehensive updates to all READMEs and MkDocs
- [x] Tests - Unit, integration, and E2E tests with 88% coverage
- [x] Tooling / CI - CircleCI configuration, Docker support

## Testing
### Commands executed
```bash
# Unit tests with coverage
uv run pytest src/ --cov=src --cov-report=term

# All CI-compatible tests
uv run pytest src/ tests/ -m "not local_credentials" --cov=src --cov-report=term

# Type checking
uv run mypy src tests

# Linting
uv run ruff check .

# Docker build and test
docker build -t mail-client-service .
docker run -d -p 8000:8000 --name mail-service mail-client-service
curl http://localhost:8000/messages
```

### Results
- [x] Tests pass (100 tests, 88% coverage)
- [x] Type checks clean
- [x] Linting clean
- [x] Manual verification - Service runs in Docker, all endpoints functional

## Quality Checklist
- [x] Contract boundaries respected; `mail_client_api` abstractions define clear interfaces
- [x] Backward compatibility confirmed - All implementations follow the same Client/Message protocol
- [x] Security / secrets handled correctly - OAuth tokens via environment variables or files, never committed
- [x] Performance impact acceptable - Lazy iteration for messages, efficient API calls
- [x] Docs updated for user-facing changes - All READMEs, MkDocs site, Quick Start guide

## Components Implemented

### 1. `mail_client_api`
- Abstract `Client` and `Message` base classes
- Factory pattern with dependency injection
- Protocol-based design for loose coupling

### 2. `gmail_client_impl`
- Gmail API integration using Google API Client
- OAuth 2.0 authentication (interactive and non-interactive modes)
- Implements `Client` interface for direct Gmail operations
- Environment variable support for CI/CD

### 3. `mail_client_service` ⭐ NEW
- **FastAPI REST API** exposing mail operations
- **ServiceClient** - HTTP client implementing `Client` interface
- **Auto-generated client** using `openapi-python-client`
- **Mock client fallback** for testing without credentials
- **Docker support** with complete containerization
- OpenAPI documentation at `/docs`

## Extra Credit Completed
- [x] **Dockerfile** - Production-ready containerization with multi-stage build
- [x] **MkDocs documentation** - Comprehensive architecture and API documentation

## Documentation Updates
- [x] Root README with CircleCI badge, Docker instructions, Quick Start
- [x] `mail_client_api/README.md` - Interface documentation
- [x] `gmail_client_impl/README.md` - Gmail integration and authentication guide
- [x] `mail_client_service/README.md` - FastAPI service, endpoints, deployment
- [x] MkDocs site - Architecture diagrams, component documentation
- [x] All documentation reflects current implementation

## Test Coverage Breakdown
```
Total Coverage: 88% (exceeds 85% requirement)
- mail_client_api: 100%
- gmail_client_impl: 85%
- mail_client_service: 100% (_impl.py), 87% (api.py)
```

## CircleCI Status
[![CircleCI](https://dl.circleci.com/status-badge/img/circleci/QJXxW5Kg3MhaRTXDr47FTf/bcb4e941-0b5f-479a-889b-9b98e69919c2/tree/dev.svg?style=shield)](https://dl.circleci.com/status-badge/redirect/circleci/QJXxW5Kg3MhaRTXDr47FTf/bcb4e941-0b5f-479a-889b-9b98e69919c2/tree/dev)

All CI tests passing with proper test markers for local vs. CI environments.

## Breaking Changes
None - This is a new implementation following the assignment specification.

## Notes for Reviewers

### Key Architecture Decisions

1. **Three-Component Design**:
   - `mail_client_api` defines contracts
   - `gmail_client_impl` provides Gmail integration
   - `mail_client_service` adds HTTP/REST layer

2. **Dependency Injection Pattern**: Factory functions are rebound at import time, allowing transparent implementation swapping

3. **Testing Strategy**:
   - Unit tests with mocks (no external dependencies)
   - Integration tests marked with `@pytest.mark.circleci`
   - E2E tests requiring credentials marked with `@pytest.mark.local_credentials`

4. **Docker Implementation**: Uses `uv` for dependency management, multi-stage build for efficiency

### Files to Review

**Core Implementation:**
- `src/mail_client_service/api.py` - FastAPI endpoints
- `src/mail_client_service/_impl.py` - ServiceClient and ServiceMessage
- `src/gmail_client_impl/src/gmail_client_impl/gmail_impl.py` - Gmail integration

**Testing:**
- `src/mail_client_service/tests/test_api_endpoints.py` - Service endpoint tests
- `tests/integration/test_service_integration.py` - Integration tests
- `tests/e2e/test_service_e2e.py` - End-to-end tests

**Documentation:**
- `README.md` - Quick Start and badges
- `docs/component.md` - Architecture overview
- `docs/api/mail_client_service.md` - Service documentation

### Demo Commands

```bash
# Run service locally
uv run uvicorn src.mail_client_service.main:app --reload

# Run in Docker
docker build -t mail-client-service .
docker run -p 8000:8000 mail-client-service

# Access endpoints
curl http://localhost:8000/
curl http://localhost:8000/messages
curl http://localhost:8000/docs  # OpenAPI documentation
```

ivanearisty committed:
Final TA Application (#2)

Squash Merging to Prevent too many commits on root.

ivanearisty committed:
Kamen requierd fixes (#4)

* millions of ruff and mypy fixes

* Major Requested Fixes

Consolidated components: Merged the message protocol and
gmail_message_impl into their client counterparts.

    Simplified architecture: Reduced the number of core packages from
four to two.

    Refactored dependency injection: Introduced an explicit register()
function for implementations.

    Added project scaffolding: Created a new root README.md and an
.env.example file.

    Overhauled documentation: Significantly condensed and rewrote all
documentation for clarity.

    Updated tests: Aligned the entire test suite with the new,
simplified structure.

* Moved from Protocols to ABC

* integration tests check v1

* integration + e2e check v2

ivanearisty committed:
Small Readme Hotfix

ranjan2601 committed:
Add E2E tests for mail_client_service and fix import issues

- Create comprehensive E2E tests for all service endpoints
- Fix import issue in ServiceClient (_impl.py) - use mail_client_api.message
- Add httpx as dev dependency for E2E testing
- E2E tests validate complete flow: HTTP → FastAPI → Real Gmail API
- Tests include: get_messages, get_message, mark_as_read, delete (skipped), error handling
- Tests require manually starting the service first (typical for E2E)

ranjan2601 committed:
Fix lint errors and add test coverage configuration

- Fix ruff lint errors with auto-fix and unsafe-fixes
- Add per-file-ignores for mail_client_service and e2e tests
- Fix import paths in service unit tests
- Add httpx to dev dependencies
- Organize imports and fix code style issues
- Current status: Lint passes, unit tests pass (coverage 64%)

tanaya09 committed:
test(service): FastAPI endpoint tests + import branches; 86% coverage (#1)

Krittin5 committed:
PR for integration tests (#2) - merging into dev

* Add integration tests for mail_client_service

- Created comprehensive integration tests in tests/integration/test_service_integration.py
- Tests verify end-to-end service call using mail_client_adapter
- Tests service integration with mocked gmail_client_impl
- All layers connected correctly: ServiceClient -> FastAPI Service -> MockClient
- Tests cover all RESTful endpoints: GET /messages, GET /messages/{id}, POST /messages/{id}/mark-as-read, DELETE /messages/{id}
- Tests handle stateful mock client behavior
- All tests marked with @pytest.mark.circleci for CI/CD compatibility
- Fixed service API import order to properly register Gmail implementation

* Fix service startup and dependency injection issues

- Add missing gmail_client_impl import in service API to register implementation
- Remove automatic service client registration from __init__.py to prevent conflicts
- Fix linting issues in integration tests (docstrings, type annotations, import order)
- Update ruff configuration to ignore E402 for integration tests
- Apply ruff formatting to all files
- All service integration tests now pass (8/8)
- Service starts successfully without NotImplementedError

ranjan2601 committed:
Clean up code: remove unnecessary comments

- Remove redundant comments from api.py, main.py
- Remove verbose import comments
- Clean up integration test comments
- Move test_service_client.py to correct location (tests/ instead of src/mail_client_service/tests/)
- Add noqa comment for DI import

ranjan2601 committed:
Fix CircleCI test failures: remove importlib.reload() calls

- Remove problematic importlib.reload() calls that fail in CI
- Modules are properly imported after sys.modules.pop()
- Add comprehensive lint suppressions for test files
- All 74 unit tests pass with 87% coverage
- All lint checks pass

ranjan2601 committed:
Add type annotations for mypy compliance

- Add type annotations to MockMessage and MockClient classes
- Add return type annotations to all FastAPI endpoints
- Add return type annotation to main.py read_root function
- Configure mypy to exclude test files
- All mypy, lint, and tests pass (87% coverage)

ranjan2601 committed:
Fix mail_client_service structure and restore coverage to 87.96%

- Moved _impl.py and generated/ from nested src/ to correct location
- Fixed test fixture to properly mock mail_client_api.Client
- Moved test_service_client.py to src/mail_client_service/tests/
- Updated test assertions to handle both mock clients
- Excluded auto-generated code from coverage reporting
- All 86 tests passing with 87.96% coverage

ranjan2601 committed:
Fix integration tests for CircleCI compatibility

- Mark test_factory_functions_work_together as local_credentials to skip in CI
- Mark test_service_to_mock_client_integration as local_credentials (requires running service)
- Mark test_service_get_specific_message_integration as local_credentials (requires running service)
- Update tests to skip gracefully when credentials unavailable
- All CircleCI tests now pass without requiring manual service startup

ranjan2601 committed:
Add Quick Start section to README

- Added setup instructions with uv commands
- Added server startup command
- Added test commands (type checking, unit tests, CI tests)
- Added Gmail authentication setup instructions

ranjan2601 committed:
Add Dockerfile and Docker instructions (extra credit)

- Created Dockerfile for containerizing FastAPI service
- Uses Python 3.11 slim base image with uv package manager
- Installs all workspace packages and dependencies
- Exposes port 8000 and runs uvicorn server
- Updated README with Docker run instructions
- Successfully tested: service runs and serves endpoints

ranjan2601 committed:
Update all documentation with CircleCI badge and comprehensive component READMEs

- Updated root README with correct CircleCI badge and links
- Updated coverage badge to 88%+
- Added Docker badge
- Completely rewrote mail_client_service README with:
  * FastAPI service documentation
  * REST API endpoints
  * Docker deployment instructions
  * ServiceClient usage examples
  * Mock client explanation
- Verified mail_client_api and gmail_client_impl READMEs are comprehensive
- All component READMEs now accurately reflect current implementation

ranjan2601 committed:
Add CONTRIBUTING.md and DESIGN.md documentation

- CONTRIBUTING.md: Comprehensive contributor guide covering architecture, testing, and development tools
- DESIGN.md: Design documentation for FastAPI service, generated client, and adapter layer

Both documents fulfill HW1 Review requirements

ranjan2601 committed:
modify design.md

ranjan2601 committed:
Update design.md

ranjan2601 committed:
Address Team 1 peer review feedback

- Add .DS_Store to .gitignore and remove all .DS_Store files
- Remove sys.path.append from api.py (uv workspace handles dependencies)
- Fix traceback usage: use logger.exception instead of import traceback in exception block
- Add mail-client-api and gmail-client-impl as dependencies in mail_client_service/pyproject.toml
- Add workspace sources for gmail-client-impl

ranjan2601 committed:
Fix ruff TRY401 error and add live demo link to README

- Remove redundant exception object from logger.exception() call
- Add live demo badge linking to deployed Swagger UI
- Add fly.io deployment links in Quick Start section

ivanearisty commented:
My commits should not be in this PR.

You should always rebase in this case since I can't tell what file changes are yours vs mine...

Maybe fast forward your main? 

Doing this: https://github.com/ranjan2601/OSPD-Fall-25/pull/3/files/39174812bc722790c2249f323e189bab0a470fb7..67405c30c2513d40481e56526b2a1fcd5157e1e2 so I can review. @adithyab-20

ranjan2601 commented:
> My commits should not be in this PR.
> 
> You should always rebase in this case since I can't tell what file changes are yours vs mine...
> 
> Maybe fast forward your main?
> 
> Doing this: https://github.com/ranjan2601/OSPD-Fall-25/pull/3/files/39174812bc722790c2249f323e189bab0a470fb7..67405c30c2513d40481e56526b2a1fcd5157e1e2 so I can review. @adithyab-20

@ivanearisty I've rebased the main branch to point to your last commit (3917481 - "Small Readme Hotfix"). 

Do you want me to create a new PR for your convinience?

ivanearisty commented on test-results/circleci/junit.xml:
JUnit is a testing framework for Java, which is confusing in a Python project that uses pytest.

Can you explain how this file was generated? 

For this project, we will rely exclusively on pytest's native reporting and coverage tools (pytest-cov). 

Please ensure your CI pipeline is configured to use pytest for all testing and reporting steps and remove this file to avoid confusion

ranjan2601 replied:
This file is auto-generated by pytest using the --junitxml flag in the CircleCI config

ivanearisty commented on src/mail_client_service/tests/test_api_endpoints.py:
Your unit tests for the API endpoints are quite complex. You're using monkeypatch to manipulate sys.modules and inject a fake mail_client_api at import time. This is a sign that the dependency injection pattern has not been fully leveraged.

A key benefit of our architecture is that it makes testing simple. Instead of patching Python's import system, you should be able to inject a mock implementation directly. For example, in your tests, you can simply replace the factory function on the imported mail_client_api module with a mock client factory before the test runs.

For example:

```python
# In your test setup:
@pytest.fixture
def mock_client():
    return MagicMock(spec=Client)

@pytest.fixture
def test_app_client(mock_client):
    from mail_client_service.main import app
    from mail_client_service.api import mail_client # The instance used by the router

    # Inject the mock directly into the running app instance
    app.dependency_overrides[mail_client] = lambda: mock_client
    yield TestClient(app)
    app.dependency_overrides.clear()

# Then your test becomes much cleaner:
def test_get_messages_success(test_app_client: TestClient, mock_client: MagicMock):
    # Arrange
    mock_message = MagicMock()
    mock_message.id = '123'
    # ... set other attributes
    mock_client.get_messages.return_value = [mock_message]

    # Act
    response = test_app_client.get("/messages")

    # Assert
    assert response.status_code == 200
    mock_client.get_messages.assert_called_once()
```

ranjan2601 replied:
Fixed. Replaced sys.modules patching with app.dependency_overrides[get_mail_client] as suggested. Tests now follow clean Arrange-Act-Assert pattern. Coverage improved to 91.31%.

ivanearisty commented on src/mail_client_service/tests/__init__.py:
inits dont belong in tests since they are not buildable

ranjan2601 replied:
removed this file

ivanearisty commented on src/mail_client_service/pyproject.toml:
You have no mypy or ruff configuration here. You must use static analyzers to objectify code quality.

ranjan2601 replied:
Fixed. Added [tool.ruff] and [tool.mypy] sections to all three new package pyproject.toml files (mail_client_service, mail_client_adapter, mail_client_service_client).

ivanearisty commented on pyproject.toml:
```diff
@@ -4,11 +4,19 @@ version = "0.1.0"
 description = "Template repository for a mail client"
 readme = "README.md"
 requires-python = ">=3.11"
-dependencies = []
```
Your pyproject.toml should not have dependencies. your inner components have dependencies.

ranjan2601 replied:
Dependencies have been removed

ivanearisty commented on pyproject.toml:
```diff
@@ -37,14 +46,20 @@ ignore = [
   "S101", # Use of `assert` detected --> Asserts are good in our opinion. Tests use asserts profusely.
 ]
-per-file-ignores = { "**/test_*.py" = ["TRY300", "BLE001", "ANN401", "SLF001", "E501", "S105", "ARG001", "ARG002", "S106", "INP001", "PLC0415"] }
+[tool.ruff.lint.per-file-ignores]
```
waaaaayyy too many ignores, every ignore should have an explanation as to why it was ignored.

ranjan2601 replied:
All ignores now have explanations. The high count is because:
1) FastAPI error handling patterns - Generic exceptions (BLE001), simplified error chains (B904), dependency injection (ARG001/ARG002), and inline error handling (TRY300/TRY301)
2) Test flexibility - Type annotations (ANN001/ANN201), magic values like HTTP status codes (PLR2004), and mocking patterns
3) Auto-generated client - Skips all checks

ivanearisty reviewed (changes requested):
TLDR, you did implement the FastAPI service, the auto-generated client, and the adapter. 

You also have included tests and documentation. 

However, the implementation deviates significantly from the established architectural patterns of the project, particularly concerning componentization, dependency management, and testing conventions. 

The current structure centralizes all new logic into a single, monolithic mail_client_service package, which conflicts with the project's core philosophy of creating small, independent, and interchangeable components.

Please refactor your project into three distinct packages within the src/ directory:

    the service: This package should only contain the FastAPI application. Its pyproject.toml should declare a dependency on mail_client_api and gmail_client_impl. This is a service, not a library.

    the adapter: This new package will contain your adapter—the concrete implementation of the mail_client_api.Client that calls the service. If I took my original main, and, instead of importing gmail_client_impl, I imported this, it should work.

	the auto generated client: This package should contain only the code generated by openapi-python-client. It will have its own pyproject.toml listing its dependencies (like httpx and attrs).

ivanearisty commented:
> > My commits should not be in this PR.
> > You should always rebase in this case since I can't tell what file changes are yours vs mine...
> > Maybe fast forward your main?
> > Doing this: https://github.com/ranjan2601/OSPD-Fall-25/pull/3/files/39174812bc722790c2249f323e189bab0a470fb7..67405c30c2513d40481e56526b2a1fcd5157e1e2 so I can review. @adithyab-20
> 
> @ivanearisty I've rebased the main branch to point to your last commit ([3917481](https://github.com/ranjan2601/OSPD-Fall-25/commit/39174812bc722790c2249f323e189bab0a470fb7) - "Small Readme Hotfix").
> 
> Do you want me to create a new PR for your convinience?

No need to create new PR, can leave as is.

ranjan2601 committed:
Refactor mail client service architecture per TA feedback

- Split monolithic mail_client_service into three packages:
  - mail_client_service: FastAPI server with dependency injection
  - mail_client_service_client: Auto-generated HTTP client
  - mail_client_adapter: Adapter implementing mail_client_api.Client

- Implement FastAPI dependency injection using Depends() pattern
- Add mock client fallback when credentials unavailable
- Remove __init__.py from test directories
- Add mypy/ruff configuration to all packages
- Document all 70+ ruff ignore rules with inline explanations
- Remove runtime dependencies from root pyproject.toml
- Fix test imports after package refactoring
- Add comprehensive tests for dependency injection and mock client
- Increase test coverage from 84.63% to 91.31%

ranjan2601 commented:
All feedback addressed. Major changes:

1. Refactored into 3 packages - Split monolithic service into:
      - mail_client_service - FastAPI server only
      - mail_client_adapter - Adapter implementing mail_client_api.Client
      - mail_client_service_client - Auto-generated HTTP client

2. Improved dependency injection - Replaced sys.modules patching with FastAPI's app.dependency_overrides pattern
3. Added static analysis - All packages now have mypy/ruff configuration
4. Documented all ignores - 70+ ruff ignores now have explanations (FastAPI patterns, test flexibility)
5. Cleaned up root config - Removed runtime dependencies from root pyproject.toml
6. Test coverage - Improved from 84.63% to 91.31%

ranjan2601 committed:
Update documentation and fix import sorting

- Update coverage badge from 88%+ to 91%+
- Fix uvicorn command to use correct module path (mail_client_service.main:app)
- Update package count from 4 to 5 (added service, client, adapter)
- Update project structure in docs to reflect 3-package refactoring
- Fix import sorting in test files (ruff auto-fix)
- Add .claude/ to .gitignore
- Update Dockerfile CMD with correct uvicorn path
